[preview] Improve URL hash synchronization

ahd44 opened the pull request:
When a preview iframe changes, we ensure that the URL hash is updated so that the URL for the preview page will generate the exact iframe shown to the user. This way, users can share URLs with each other (or bookmark a page and return to it) and ensure they will see the same preview.*

Adds explicit parameter checking when interacting with search or hash parameters and fixes type declarations to match usage. Reduces history pollution by using `replaceState` instead of `pushState` unless explicitly needed/wanted.

*There are a handful of cases where there isn't true. Two cases are: 1) parameters included in iframe URL but not respected on load (e.g. NetCDF previews don't respect the `color` parameter); 2) interactive options that are not included in iframe URL right now, but theoretically could be (e.g. changes to the "range" options in NetCDF previews, or zooming in or out on a time series Dygraph); in this case, we would need to decide which ones should be included.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209593710989303

ahd44 commented on src/main/js/portal/main/models/Preview.ts:
```diff
@@ -18,50 +18,95 @@ export function previewVarCompare(po1: PreviewOption, po2: PreviewOption): numbe
+	'y1',
+	'map'
+]
+
```
I don't really like how I did this but for good sanitization it seems necessary. There are ways to "extract" the key names from the type, but they are all very convoluted and don't have good run-time safety. This feels like it's duplication (and a pain to keep in sync), but matches what we are doing in `State.ts` (for `hashKeys` and `jsonHashState`).

The idea, though, is to allow only these keys through any kind of URL hash or query. This way we can check if it is an "allowed" key before passing it into anything, avoiding collisions of other objects and ensuring that we are not passing any unexpected parameters.

Probably more important within `State.ts` itself, since those properties get merged with the State and it is possible to overwrite variables without the checking. Also, I called my functions "sanitization" but I don't know if they really are, more of just... filtering, I guess.

ahd44 replied:
Sorry, I'm not understanding this one. The position of the objects/keys isn't actually important, they can be in any order.

jonathanthiry replied:
Good job, that looks good!

ahd44 commented on src/main/js/portal/main/models/Preview.ts:
```diff
@@ -121,20 +166,20 @@ export default class Preview {
 		this.pids = pids;
+		this.previewSettings = previewSettings ?? {};
 		return this;
 	}
```
I think this could be a static function (more like a factory function) as, in practice, it is always called on a newly constructed `Preview`. It would instead construct a new Preview, alter those variables, and then return it.

ahd44 commented on src/main/js/portal/main/models/Preview.ts:
```diff
@@ -207,5 +252,5 @@ export function batchPreviewAvailability(
 			return noPreview("You can only batch-preview plain time series data objects")
 	}
-	return {previewType: previewTypes.values().next().value}
+	return {previewType: previewTypes.values().next().value!}
```
Was getting a TypeScript warning about this code, so now explicitly telling TypeScript compiler that there will be a value here.

ahd44 commented on src/main/js/portal/main/models/State.ts:
```diff
@@ -39,12 +39,13 @@ const hashKeys = [
+	'searchOptions',
+	'mapProps',
 	'itemsToAddToCart'
 ];
```
`hashKeys` was reorganized to match `JsonHashState` object order below. I would also suggest moving them next to each other to make it more explicit that they are really referring to the same thing, but that would be a "big change" in the git diff, so refrained from doing so.

Note that `yAxis` and `y2Axis` are retained here (and in `JsonHashState`) to allow for previous links to continue working. These are explicitly converted into `previewSettings.y` and `previewSettings.y2` (in `sanitizeJsonHash`)

ahd44 commented on src/main/js/portal/main/models/State.ts:
```diff
@@ -346,14 +347,41 @@ type JsonHashState = {
+	searchOptions?: SearchOptions
+	mapProps?: MapProps
+	itemsToAddToCart?: Sha256Str[]
 }
```
The only actually "new" addition here is `previewSettings`; the rest were being used at run-time but not explicitly declared in TypeScript. Now, with the added `sanitizeJsonHash` function below, we are using only the `hashKeys`, which match the keys of `JsonHashState`, to ensure we are not overwriting our state or including parameters that we do not expect.

ahd44 commented on src/main/js/portal/main/models/State.ts:
```diff
@@ -416,18 +443,15 @@ function getCurrentHash(){
+		portalHistoryState.replaceState(serialize(state), window.location.href.split('#')[0] + newHash).then(
+			_ => _,
+			reason => console.log(`Failed to add value to indexed database because ${reason}`)
+		);
```
`portalHistoryState.pushState` adds an entry to the session history stack, so goal was to change it to `portalHistoryState.replaceState` unless we wanted to actively add a new entry. `window.location.hash`, however, was actually responsible for most of the "pollution" in the history stack; it kind of silently adds an entry to it. Changing both cases to use `replaceState`, then, fixes this problem.

ahd44 commented on src/main/js/portal/main/models/Preview.ts:
```diff
@@ -18,50 +18,95 @@ export function previewVarCompare(po1: PreviewOption, po2: PreviewOption): numbe
 	return po1.varTitle.localeCompare(po2.varTitle)
 }
 
+export type PreviewSettings = {
```
These are all strings because they represent the "stringified version" of these variables, since they are URL parameters within the iframe. Technically these could be more specific or defined (which would then require some typecasting later) but in the context of these canonically representing how one can re-create a persistent iframe URL, I think string is appropriate. (e.g., `type?: 'scatter' | 'line; img?: number; gamma?: number; center: string | {lat: number, lng: number};` -- in some cases where the param is used for multiple previews it can be a bit messier.

I also thought about having different types for each `PreviewType`, and then having a composite union type of `PreviewSettings`, but this seemed like too much overhead/maintenance to me.

ahd44 commented on src/main/js/map-graph/main/store.js:
```diff
@@ -57,7 +57,7 @@ export const updateURL = store => () => {
 			window.top.postMessage(newURL, '*');
 		} else {
-			history.pushState({urlPath: newURL}, "", newURL);
+			history.replaceState({urlPath: newURL}, "", newURL);
```
You're correct; if we are actually changing context, we should use `pushState`. The issue is that, currently, when one interacts with the Preview it updates the history, but we have found that most people will view a preview, tinker with it a bit, then try to go back to the data object or search results. Since we are making new history states for every change to the preview, this does not work, and you end up needing to hit the button several times. You can see this at https://data.icos-cp.eu/portal/#%7B%22route%22:%22preview%22,%22preview%22:%5B%22dy7z7hIUZG-bJlBehgJMfR2l%22%5D%7D if you change the select drop-downs.

This specific change is to use `replaceState` in full-screen mode, which I also think would be a better user experience. I think it's unexpected to interact with a graph through drop-down menus and then have the back button undo those changes instead of taking you back to the referring page.

(Additionally, this doesn't work well right now regardless; it is pretty inconsistent because the `pushState` is called too many times and then the select drop-downs don't reflect what is actually in the graph. This could be fixed, of course, if we wanted to support that behavior.)

ahd44 commented on src/main/js/portal/main/components/preview/PreviewSelfContained.tsx:
```diff
@@ -2,21 +2,23 @@ import React, {ChangeEvent, useState, useEffect, useRef, useCallback} from 'reac
 
 	const [height, setHeight] = useState<number>(() => getInitialHeight(preview.type));
-	const src = useRef<string>(preview.type ? getPreviewIframeUrl(preview.type, preview.item) : "");
+	const src = useRef<string>(preview.type ? getPreviewIframeUrl(preview.type, preview.item, previewSettings) : "");
```
Good point... `preview.type` can be `undefined` here, so I could make `getPreviewIframeUrl` expect either `PreviewType` or `undefined` and then return an empty string if `previewType` is `undefined`, but that feels similarly unwieldy to me.

I think this "ternary type guard" is a pretty common pattern in TypeScript for dealing with potentially undefined values... otherwise you need to account for them in your function type signature. So either way you are adding overhead, just depends on where you want it.

ahd44 commented on src/main/js/portal/main/components/preview/PreviewSelfContained.tsx:
```diff
@@ -75,15 +77,29 @@ function shouldUpdateHeight(previewType?: PreviewType): boolean {
 
-function getPreviewIframeUrl(previewType: PreviewType, item: CartItem): UrlStr {
+function settingsToQueryParams(previewSettings: PreviewSettings): string {
+	let queryParams = Object.keys(previewSettings).map((key) => `${key}=${previewSettings[key as keyof PreviewSettings]}`).join("&");
```
Going to change to:

```
function settingsToQueryParams(previewSettings: PreviewSettings): string {
	const remapKey = (key: string) => `${key}=${previewSettings[key as keyof PreviewSettings]}`;
	return Object.keys(previewSettings).map(remapKey).join("&");
}
```

ahd44 commented on src/main/js/portal/main/components/preview/PreviewSelfContained.tsx:
```diff
@@ -75,15 +77,29 @@ function shouldUpdateHeight(previewType?: PreviewType): boolean {
 	const hashId = getLastSegmentInUrl(item.dobj);
-	if (previewType === config.PHENOCAM) {
-		return `${iFrameBaseUrl}?objId=${hashId}`;
+	switch (previewType) {
+		case config.PHENOCAM:
+			return `${iFrameBaseUrl}?objId=${hashId}${previewSettings.img ? `&img=${previewSettings.img}` : ''}`;
+		case config.MAPGRAPH:
+			return `${iFrameBaseUrl}${hashId}#${settingsToHash(hashId, previewSettings)}`;
+		case config.NETCDF:
+			return `${iFrameBaseUrl}${hashId}?${settingsToQueryParams(previewSettings)}`;
 	}
 	return iFrameBaseUrl + hashId;
```
`previewType` should always be defined due to the type guard (discussed above), or we would be type guarding it here before using it.

> It looks to me, right now, as if adding a new type would result in easily forgetting to update this function, and in turn getting a valid but functionally incorrect result?

Yes, it would be possible for that to be the case. Unfortunately since the different previews were developed at different times/by different developers, they don't currently share the same URL format. Even the two that both use query parameters do it differently (`config.PHENOCAM` and `config.NETCDF`). Ideally they would follow the same usage pattern and then we would not need to specifically translate the previewSettings to match the format, and adding a new preview type would work more seamlessly.

The switch should only be reached on the initial page load, so returning a "base URL" that is likely correct without any options applied should be OK, but failing loudly would make it more likely for a developer to notice that the new preview will not be responsive to hash parameters on initial load. I would prefer for there to be one "default format", and then default to that, e.g. replace `case config.NETCDF:` with `default:` (and I think I actually wrote that at some point before changing it).

So... I think we could go down any of those paths.

Adding a new preview type currently would involve changing a few files right now. This file, `PreviewSelfContained.tsx`. The `Preview.ts` model would need its `PreviewSettings` type (and its keys) changed as well, and `config.ts` would need to be edited to include the new type. But that should be it as far as I can tell (except for the logic of associating a preview type with a specific file format/dataset/etc, which I have not at all looked into). It may be possible to move the `previewType` specific actions into the `config.ts` file, but we would need to convert the previews to use the same URL parameter strategy before doing so.

ahd44 commented on src/main/js/portal/main/components/preview/PreviewTimeSerie.tsx:
```diff
@@ -171,7 +162,8 @@ export default class PreviewTimeSerie extends Component<OurProps, OurState> {
 		const legendLabelsY2Params = y2Axis && legendLabels ? `&legendLabelsY2=${legendLabels}` : '';
-		const iframeUrl = `${window.document.location.protocol}//${window.document.location.host}${iFrameBaseUrl}?objId=${objIds}&x=${xAxis}${yParam}${y2Param}&type=${type}&linking=${linking}${legendLabelsParams}${legendLabelsY2Params}`;
+		const typeParam = type ? `&type=${type}` : '';
+		const iframeUrl = `${window.document.location.protocol}//${window.document.location.host}${iFrameBaseUrl}?objId=${objIds}&x=${xAxis}${yParam}${y2Param}${typeParam}&linking=${linking}${legendLabelsParams}${legendLabelsY2Params}`;
```
Since we never reuse these variables, I opted to create an `Array` which is then joined together:

```
		const iframeParams: Array<string> = [`?objId=${objIds}`];
		iframeParams.push(`&x=${xAxis}`);
		iframeParams.push(yAxis ? `&y=${yAxis}` : '');
		iframeParams.push(y2Axis ? `&y2=${y2Axis}` : '');
		iframeParams.push(`&linking=${linking}`);
		iframeParams.push(legendLabels ? `&legendLabels=${legendLabels}` : '');
		iframeParams.push(y2Axis && legendLabels ? `&legendLabelsY2=${legendLabels}` : '');
		iframeParams.push(type ? `&type=${type}` : '');

		const iframeUrl = `${window.document.location.protocol}//${window.document.location.host}`
							+ `${iFrameBaseUrl}${iframeParams.join('')}`;
```

Thought about doing something similar to `settingsToQueryParams` from `PreviewSelfContained.tsx` but the existence checking makes things more complex; this approach makes it clear how we are constructing the URL while reducing namespace clutter, I think, and having a shorter/easier to understand assignment to `iframeUrl`.

ahd44 commented on src/main/js/portal/main/models/CartItem.ts:
```diff
@@ -57,15 +57,21 @@ export default class CartItem {
-		const search = url.split('?').pop() || '';
-		const searchStr = search.replace(/^\?/, '');
-		const keyValpairs = searchStr.split('&');
+		if (url.includes("?")) {
+			const searchStr = url.split('?').pop() || '';
+			const keyValpairs = searchStr.split('&');
+
+			return keyValpairs.reduce<IdxSig>((acc, curr) => {
+				const p = curr.split('=');
+				acc[p[0]] = p[1];
+				return acc;
+			}, {});
+		}
```
Good catch; I was mostly keeping the previous code, so I didn't really think about how it could be improved.

But this is much nicer:
```
const searchParams = Object.fromEntries(new URL(url).searchParams);
if (Object.keys(searchParams).length !== 0) {
	return searchParams;
}
```

ahd44 commented on src/main/js/portal/main/models/CartItem.ts:
```diff
@@ -57,15 +57,21 @@ export default class CartItem {
-			acc[p[0]] = p[1];
-			return acc;
-		}, {});
+		if (url.includes("#")) {
+			const hashStr = url.split("#").pop() || '';
+			return JSON.parse(decodeURIComponent(hashStr));
+		}
```
The previous change led nicely into this one, which avoids both constructions:

```
if (webUrl.hash) {
	return JSON.parse(decodeURIComponent(webUrl.hash));
}
```

jonathanthiry commented on src/main/js/portal/main/components/preview/PreviewTimeSerie.tsx:
```diff
@@ -158,15 +158,17 @@ export default class PreviewTimeSerie extends Component<OurProps, OurState> {
-
-		if (!preview)
-			return null;
+		const iframeParams = [`?objId=${objIds}`];
+		iframeParams.push(`&x=${xAxis}`);
+		iframeParams.push(yAxis ? `&y=${yAxis}` : '');
+		iframeParams.push(y2Axis ? `&y2=${y2Axis}` : '');
+		iframeParams.push(`&linking=${linking}`);
+		iframeParams.push(legendLabels ? `&legendLabels=${legendLabels}` : '');
+		iframeParams.push(y2Axis && legendLabels ? `&legendLabelsY2=${legendLabels}` : '');
+		iframeParams.push(type ? `&type=${type}` : '');
```
Much better! I'm wondering if it could be worth using `concat` or the spread operator to remove the `iframeParams.push` repetition.

ahd44 replied:
Have rewritten to directly define the array with those values; that is definitely cleaner and more efficient to the way I originally wrote it.

```
const iframeParams = [`?objId=${objIds}`,
	`&x=${xAxis}`,
	yAxis ? `&y=${yAxis}` : '',
	y2Axis ? `&y2=${y2Axis}` : '',
	`&linking=${linking}`,
	legendLabels ? `&legendLabels=${legendLabels}` : '',
	y2Axis && legendLabels ? `&legendLabelsY2=${legendLabels}` : '',
	type ? `&type=${type}` : '',
];
```

jonathanthiry commented on src/main/js/portal/main/components/preview/PreviewTimeSerie.tsx:
```diff
@@ -158,15 +158,17 @@ export default class PreviewTimeSerie extends Component<OurProps, OurState> {
+		iframeParams.push(y2Axis && legendLabels ? `&legendLabelsY2=${legendLabels}` : '');
+		iframeParams.push(type ? `&type=${type}` : '');
+
+		const iframeUrl = `${window.document.location.protocol}//${window.document.location.host}`
+							+ `${iFrameBaseUrl}${iframeParams.join('')}`;
```
Would it be worth splitting it in 3-4 lines?
We usually use spaces for alignment.

jonathanthiry reviewed (approved):
So many nice improvements! I didn't test everything but it looks well done.

ggVGc commented on src/main/js/portal/main/components/preview/PreviewTimeSerie.tsx:
```diff
@@ -167,14 +158,20 @@ export default class PreviewTimeSerie extends Component<OurProps, OurState> {
-
-		if (!preview)
-			return null;
+		const iframeParams = [`?objId=${objIds}`,
+			`&x=${xAxis}`,
+			yAxis ? `&y=${yAxis}` : '',
+			y2Axis ? `&y2=${y2Axis}` : '',
+			`&linking=${linking}`,
+			legendLabels ? `&legendLabels=${legendLabels}` : '',
+			y2Axis && legendLabels ? `&legendLabelsY2=${legendLabels}` : '',
+			type ? `&type=${type}` : '',
+		];
+
+		const iframeUrl = `${window.document.location.protocol}//`
+			+ `${window.document.location.host}`
+			+ `${iFrameBaseUrl}`
+			+ `${iframeParams.join('')}`;
```
Sorry to comment on this again, but I read your comments and had some more thoughts on this. Since there's been some efforts improving this code, I think we can go even further. I believe the following suggestion is equivalent, but I didn't test it fully.
`iframeUrl' is now a 'URL', so if it's important that it's a string, the two usages further down needs a `toString()` added.

```suggestion
		let iframeUrl = new URL(iFrameBaseUrl, window.document.location.origin)

		iframeUrl.searchParams = new URLSearchParams([
      ['objId', objIds], 
      ['type', type], 
      ['linking', linking],
      ['x', xAxis]
    ])

    if (yAxis) {
      iframeUrl.searchParams.append('y', yAxis)
    }

    if (y2Axis) {
      iframeUrl.searchParams.append('y2', y2Axis)
    }

    if (legendLabels) {
      iframeUrl.searchParams.append('legendLabels', legendLabels)
    }

    if (y2Axis && legendLabels) {
      iframeUrl.searchParams.append('legendLabelsY2', legendLabelsY2)
    }
```

ggVGc replied:
Yeah, wow, I didn't know the `URLSearchParams` had such overhead. I thought it was basically a `Map`.
Anyway, my last contribution is this, which in my test is the fastest :)
My main input, though, was supposed to be that I think it's clearer to explicitly set the conditional things in if-blocks (as in my first suggestion and this one), rather than setting fields to null or empty and then filtering them later. 

```js
    let params = {
      objId: objIds,
      x: xAxis,
      linking,
      type
    };

    if (yAxis) {
      params.y = yAxis
    }

    if (y2Axis) {
      params.y2 = y2Axis
    }

    if (legendLabels) {
      params.legendLabels = legendLabels
    }

    if (y2Axis && legendLabels) {
      params.legendLabelsY2 = legendLabelsY2
    }

    const iframeParamsStr = Object.entries(params)
      .map((param) => param.join('='))
      .join("&");
```

ahd44 replied:
Ah, I see. It's rather verbose but definitely more readable and clear. The type definition ends up a bit odd (`let params: PreviewSettings & { objId: string }`), but it is definitely a nice improvement, and I've added it in.

ahd44 commented:
Had to refactor `PreviewTimeSerie.tsx` into a function-based component in order to use a reference for `initialIframeUrl`. This allows us to avoid re-rendering the entire iframe, making the preview both more efficient/performant and correcting for history pollution via the `src` of the iframe changing. I also found a bug where going from search directly to preview did not `pushState`; apparently we were relying on the hash changing as our "should push state" metric, which is now changed to detecting when the route changes (which is stored in the hash, unless it is empty when means "search").

Also added some URI encoding where it was missing; I am not sure how I missed it before, but it is working appropriately now. I will continue to test this and ensure there are no other bugs, but it seems much better to me now.
